keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,401 @@
+   using the efficient [Clone()](https://pkg.go.dev/github.com/google/btree#BTree.Clone) method.
+   This clone is a lazy copy, only duplicating necessary nodes, resulting in
+   minimal overhead and effectively is container of pointers in-memory and not actual copies
+   of the B-tree nodes. Since we do not edit event histories in the watchCache, a snapshot
+   is treated as a read-only entity and its creation is cheap (creating only pointers in memory).
+2. **Snapshot Storage:** The snapshot will be stored in memory, keyed by
+   resourceVersion the watchCache has propagated upto when the snapshot was created. It is also this
+   resourceVersion that is returned as part of the continuation token in the response. 
+3. **Serving Subsequent Pages:** When a subsequent request with a continue token
+   arrives, the API server will:
+  - Extract the resourceVersion from the continue token. 
+  - Since resourceVersions provide a global logical clock sequencing all events in the cluster, a snapshot
+    of the watchCache for this resourceVersion is retrieved using the resourceVersion as the key.
+  - The corresponding snapshot may not be present in the following 2 scenarios at an API Server:
+    - Snapshot has been cleaned up due to the 75s TTL (see below).
+    - In a HA setup, the request goes to an API Server that did not get the initial LIST request (and hence does not have the in-memory snapshot created as described in (1)).
+4. **Snapshot Cleanup:** Snapshots will be subject to a Time-To-Live (TTL)
+   mechanism. We will reuse the existing watch event cleanup logic, which has a
+   75s TTL. This ensures that snapshots don't accumulate indefinitely.
+
+As still some pagination requests will be delegated to etcd, we will monitor the
+success rate by measuring the pagination cache hit vs miss ratio.
+
+Consideration: Should we start respecting the limit parameter?
+
+### Risks and Mitigations
+
+#### Client setting limit while not supporting pagination
+
+#### Memory overhead
+
+No, B-tree only store pointers the actual objects, not the object themselves.
+The objects are already cached to serve watch, so it should only add a small
+overhead for the B-tree structure itself, which is negligible compared to the
+size of the cached objects.
+
+#### Pagination request hitting another apiserver
+
+When connecting directly expectation is that a client will stay connected to the
+same apiserver. What happens when apiserver is behind loadbalancer?
+
+For setups with L4 loadbalancer apiserver can be configured with Goaway, which
+requests client reconnects periodically, however per request probability should
+be configured around 0.1%.
+
+For L7 loadbalancer the default algorithm usually is round-robin. For most LBs
+it should be possible to switch the algorithm to be based on source IP hash.
+Even if that is not possible, stored snapshots will never be used and user will
+not be able to benefit from the feature.
```

Comment:
I'm not sure we can expect providers to change their LB configuration...

Reply:
My main point here was that we expect only minimal overhead, while users of L7 LB can can opt in via a common configuration option.